src/hotspot/share/oops/objArrayKlass.cpp

```diff
@@ -432,6 +433,14 @@ PackageEntry* ObjArrayKlass::package() const {
   return bottom_klass()->package();
 }
 
+#if INCLUDE_CDS_JAVA_HEAP
+void ObjArrayKlass::deallocate_contents(ClassLoaderData* loader_data) {
+  if (DumpSharedSpaces) {
+    HeapShared::remove_scratch_objects(this);
+  }
+}
+#endif
```

Comment:
When will the above function be called?

From `ClassLoaderData::free_deallocate_list()`

```
      // Cast them so they can be used by the template function.
      if (m->is_method()) {
        MetadataFactory::free_metadata(this, (Method*)m);
      } else if (m->is_constantPool()) {
        MetadataFactory::free_metadata(this, (ConstantPool*)m);
      } else if (m->is_klass()) {
        MetadataFactory::free_metadata(this, (InstanceKlass*)m);
      } else {
        ShouldNotReachHere();
      }
```
`free_metadata` will call `deallocate_contents` but I don't see how it will call the` ObjectArrayKlass` version.

Reply:
Thanks for finding this. `<T>::free_deallocate_list` is called only by the template function `MetadataFactory::free_metadata`. I removed `ObjArrayKlass::deallocate_contents()` and HotSpot still builds. This means nobody uses `free_metadata` with the `ObjArrayKlass` type.